File: ai-service-python/app/models/trade_signal.py
from datetime import datetime
from decimal import Decimal

from sqlalchemy import String, SmallInteger, Numeric, Text, DateTime, ForeignKey, CheckConstraint
from sqlalchemy.dialects.postgresql import UUID, ARRAY
from sqlalchemy.orm import Mapped, mapped_column

from app.models.base import Base, UUIDMixin, TimestampMixin


class TradeSignal(UUIDMixin, TimestampMixin, Base):
    __tablename__ = "trade_signals"
    __table_args__ = (
        CheckConstraint("signal_type IN ('BUY', 'SELL', 'HOLD')", name="chk_signal_type"),
        CheckConstraint("confidence >= 0 AND confidence <= 100", name="chk_confidence"),
    )

    user_id: Mapped[str] = mapped_column(UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False)
    fund_id: Mapped[str] = mapped_column(UUID(as_uuid=True), ForeignKey("funds.id", ondelete="CASCADE"), nullable=False)
    signal_type: Mapped[str] = mapped_column(String(8), nullable=False)
    strategy: Mapped[str] = mapped_column(String(64), nullable=False)
    confidence: Mapped[Decimal] = mapped_column(Numeric(5, 2), nullable=False)
    reason: Mapped[str | None] = mapped_column(Text)
    related_news_ids: Mapped[list[str]] = mapped_column(ARRAY(UUID(as_uuid=True)), default=list)
    target_amount: Mapped[Decimal | None] = mapped_column(Numeric(14, 2))
    execute_before: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    status: Mapped[int] = mapped_column(SmallInteger, default=0)
    executed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))

File: backend-go/internal/model/trade_signal.go
package model

import (
	"time"

	"github.com/shopspring/decimal"
)

// TradeSignal 交易信号模型
type TradeSignal struct {
	ID             string           `gorm:"type:char(36);primaryKey" json:"id"`
	UserID         string           `gorm:"type:char(36);index" json:"user_id"`
	FundID         string           `gorm:"type:char(36);index" json:"fund_id"`
	SignalType     string           `gorm:"type:varchar(8)" json:"signal_type"`
	Strategy       string           `gorm:"type:varchar(64)" json:"strategy"`
	Confidence     decimal.Decimal  `gorm:"type:decimal(5,2)" json:"confidence"`
	Reason         *string          `gorm:"type:text" json:"reason,omitempty"`
	RelatedNewsIDs StringArray      `gorm:"type:json;default:'[]'" json:"related_news_ids"`
	TargetAmount   *decimal.Decimal `gorm:"type:decimal(14,2)" json:"target_amount,omitempty"`
	ExecuteBefore  *time.Time       `json:"execute_before,omitempty"`
	Status         int16            `gorm:"default:0" json:"status"`
	ExecutedAt     *time.Time       `json:"executed_at,omitempty"`
	CreatedAt      time.Time        `json:"created_at"`
	UpdatedAt      time.Time        `json:"updated_at"`

	User User `gorm:"foreignKey:UserID" json:"user,omitempty"`
	Fund Fund `gorm:"foreignKey:FundID" json:"fund,omitempty"`
}

func (TradeSignal) TableName() string { return "trade_signals" }
